README.md
# Sha#Bang! Modules

A collection of innovative, probabilistic, generative, and rhythmically complex modules for [VCVRack](https://vcvrack.com/). Created by [Jeremy Muller](http://jeremymuller.com/). If you enjoy using these modules and/or have made money with these, please consider [sponsoring](https://github.com/sponsors/jeremymuller) this project or giving a donation through [PayPal](https://www.paypal.com/paypalme/jeremysmuller). Huge shoutout to **Andras Szabo** for being the very first sponsor of Sha#Bang! Modules! Thank you :pray:

Add these modules to VCVRack through the [Rack library](https://library.vcvrack.com/?brand=Sha%23Bang!%20Modules).

![modulesall](/docs/Modules_ALL5.gif)

## Contents:

* [Collider](#collider) :tv:
* [Cosmosis](#cosmosis) :tv:
* [Neutrinode](#neutrinode) :tv:
* [Orbitones](#orbitones) :tv:
* [Photron](#photron--photron-panel--photron-strip)
* [PhotronPanel](#photron--photron-panel--photron-strip)
* [PhotronStrip](#photron--photron-panel--photron-strip)
* [PolyrhythmClock](#polyrhythm-clock) :tv:
* [QubitCrusher](#qubit-crusher)
* [RandGates](#randgates) :tv:
* [RandRoute](#randroute)
* [StochSeq](#stochseq) :tv:
* [StochSeq4](#stochseq4) :tv:
* [StochSeq4X](#stochseq4x)
* [StochSeqGrid](#stochseqgrid) :tv:
* [Talea](#talea) :tv:

## Video playlists:

* [All tutorials](https://youtube.com/playlist?list=PLqgOVd1WHf5vFl7vLlk3UglMzuUWLnZng)
* [VCV Rack patches that use Sha#BANG! modules](https://youtube.com/playlist?list=PLqgOVd1WHf5t23rEmbKc0bABRZTvfx69n)

---

### Collider

![Collider](/docs/Collider.png)

*A physical model of various of shakers and wind chimes (maracas, sleigh bells, bamboo chimes, metallic chimes, etc).*

Watch the tutorial:

[![Collider_video](docs/Collider-video.png)](https://youtu.be/nC_NozyveTc "Collider tutorial")

##### RIGHT-CLICK MENU:
- Polyphony.
##### BUTTON:
- `SHAKE` shakes the particles. Hold down to continuously shake.
##### INPUTS:
- `SHAKE` gate controls the shake.
- `V/OCT` center frequency of the particles.
- `SPREAD` control voltage determines the amount of spread from the center frequency.
- `VEL` control voltage for the initial shake energy.
- `PARTICLES` control voltage for the number of particles. More particles = faster decay in system energy, less particles = slower decay.
##### KNOBS:
- `C` center frequency of the particles in Hz.
- `SPREAD` sets the amount of spread from the center frequency.
- `RND` sets the amount of random frequencydeviation on each particle collision.
- `PARTICLES` sets the number of particles. More particles = faster decay in system energy, less particles = slower decay.
##### OUTPUTS:
- `V/OCT` outputs ±5 volts.
- `GATE` outputs pulses.
- `VEL` outputs velocity of entire system.

---

### Cosmosis

![Cosmosis](/docs/Cosmosis.png)

*A continuous sequencer based on constellations where time can move left->right, right->left, up->down, and down->up.*

Click anywhere to add/drag new stars. To remove stars click/drag them out of the dark display area.

Watch the tutorial:

[![Cosmosis_video](docs/Cosmosis-video.png)](https://youtu.be/isZ22FNc9eI "Cosmosis tutorial")

##### RIGHT-CLICK MENU:
- Polyphony.
##### BUTTONS:
- `PLAY` starts the sequencer.
- `CLR` removes all stars.
- `POS` randomizes positions of stars.
- `RAD` randomizes the radii of stars.
##### INPUTS:
- `PLAY` trigger starts/stops the sequencer.
- `SPEED` control voltage manipulates the current speed.
- `RESET` trigger resets the current sequence position.
- Root note input will accept V/OCT to set the root note.
- `POS` accepts gate that triggers randomizing positions of stars.
- `RAD` accepts gate that triggers randomizing radii of stars.
##### KNOBS:
- `SPEED` tempo of sequencer.
- `PATT` sets the constellation pattern.
- Two knobs control the root note and the scale. The Messiaen modes are based on the wonderful French composer, Olivier Messiaen and his [modes of limited transposition](https://en.wikipedia.org/wiki/Mode_of_limited_transposition) found in his book *The Technique of My Musical Language*.
- Chords like MM7, Mm7, mm7, etc. are based on [seventh chords](https://en.wikipedia.org/wiki/Seventh_chord).
- `OCT` controls the octaves of individual sequencers.
- `MODE` sets which direction time will traverse the stars.
##### SWITCH:
- `PITCH` left side will take pitch from the radius of particles. The right side will take pitch based on the X or Y position (width or height respectively) in the display. If the `MODE` is set to either Blue or Red, in which case time travels down or up, pitch is taken from the horizontal position of the stars. The left switch is good if you'd like to generate randomness in the pitches, however, if you want more control over the pitch then select the right switch.
##### OUTPUTS:
- `GATES` output pulses.
- `V/OCT` outputs ±5 volts.

---

### Neutrinode

![Neutrinode](/docs/Neutrinode.png)

*A unique visual-based sequencer where time moves from the center of the 4 independent nodes out to the connected particles. Can generate interesting rhythmic textures and can be animated so that the textures change over time.*

Click on nodes to position them. Click anywhere else to add/drag new particles. To remove particles click/drag them out of the dark display area.

Watch the tutorial:

[![Neutrinode_video](docs/Neutrinode-video.png)](https://youtu.be/WVPZnpNk_UE "Neutrinode tutorial")

##### RIGHT-CLICK MENU:
- Continuous play or one-shot mode (useful for synchronization). These options only affect the `PLAY` button/input.
- Collisions are turned on or off for the nodes.
- Polyphony.
##### BUTTONS:
- `PLAY` turns on/off all nodes.
- `MOVE` each node will randomly move around the dispay area.
- `CLR` removes all particles.
- `ON` turns on/off individual nodes.
##### INPUTS:
- `PLAY` trigger starts/stops the generator.
- `BPM` control voltage manipulates the current bpm.
- `MOVE` trigger turns on/off the random movement of nodes.
- Root note input will accept V/OCT to set the root note.
##### KNOBS:
- `BPM` tempo of all nodes.
- `SPEED` is the velocity of movement of nodes when `MOVE` is switched on.
- Two knobs control the root note and the scale. The Messiaen modes are based on the wonderful French composer, Olivier Messiaen and his [modes of limited transposition](https://en.wikipedia.org/wiki/Mode_of_limited_transposition) found in his book *The Technique of My Musical Language*.
- Chords like MM7, Mm7, mm7, etc. are based on [seventh chords](https://en.wikipedia.org/wiki/Seventh_chord).
- `OCT` controls the octaves of individual nodes.
##### SWITCHES:
- `PITCH` left side will take pitch from the radius of particles. The right side will take pitch based on the Y position (height) in the display. The left switch is good if you'd like to generate randomness in the pitches, however, if you want more control over the pitch then select the right switch.
##### OUTPUTS:
- `GATES` and `V/OCT` output only from the corresponding node color.
- `ALL` outputs all nodes.

*:warning: There are a maximum of 16 particles (for 16 polyphonic channels) per node. However, when using the `ALL` outputs, there is potential to reach that maximum (16 x 4 = 64) in which case it does voice stealing.*


---

### Orbitones

![Orbitones](/docs/Orbitones.gif)

*Physics-based animated LFO with 4 independent attractors.*

Click on the attractors to move position them where you want. Click anywhere else to add particles.

Watch the tutorial:

[![Orbitones_video](docs/Orbitones-video.png)](https://youtu.be/x7X0--AugRU "Orbitones tutorial")

##### RIGHT-CLICK MENU:
- Polyphony.
- Particle trails:
  - off
  - white
  - red/blue shift (based on velocity of particles)
- Particle boundaries:
  - on (particles are bound by the display area and will bounce off edges)
  - off (particles are able to leave the display area)
##### BUTTONS:
- `RMV` removes one particle.
- `CLR` clears all particles.
- `MOVE` trigger turns on/off the random movement of attractors.
##### KNOBS:
- `ON` turns on/off individual attractors.
- `G` scales the individual attractors' gravity.
- `G` (big knob) main gravity control for all attractors.
- `OFFSET` offsets the voltage output.
##### INPUTS:
- `MOVE` trigger turns on/off the random movement of attractors.
- `G` takes a CV using this formula: `G` * 2<sup>V</sup>.
##### OUTPUTS:
- `MONO OUTS`:
  - `AVG` average x/y voltage of all particles.
  - `MAX` maximum x/y voltage of all particles.
  - `MIN` minimum x/y voltage of all particles.
- `POLY OUTS`:
  - `X Y` positions of particles correspond to ±5 volts.
  - `-X -Y` opposite positions of particles correspond to ±5 volts.
  - `velX velY` velocities of particles correspond to ±5 volts.

---

### Photron | Photron Panel | Photron Strip

![Photron](/docs/Photron-Panel-Strip.png)

*Animated visualizers that uses color flocking based on the Craig Reynolds boids flocking algorithm.*

Placing modules adjacent to each other will cause them to blend colors. `PhotronPanel` and `PhotronStrip` will work together, however `Photron` only will blend with another `Photron`.

#### Photron

##### RIGHT-CLICK MENU:
- `Processing rate` (for those with slower CPUs). Keep in mind, if you slow the processing rate down, it'll help your CPU but the animation will also slow down.
- `Lissajous mode` on or off.
- `Pattern` select from a list of 8-bit retro characters and others.
- `Lock pattern` lock the pattern or let it blend.
##### BUTTONS:
- Waveform mode: lines, blocks, or off.
- Background mode: color, black & white, or black.
##### KNOBS:
- Purple knobs adjust the X and Y offset of the waveform(s).
- Blue knobs adjust the X and Y scaling of the waveform(s).
##### INPUTS:
- TOP 4 inputs (zero volts is default values):
  - CV controls the separation of colors between adjacent blocks. Higher voltage = more separation.
  - CV controls the alignment of block colors which is their rate of change. Higher voltage = more alignment and thus their color velocities will become the same.
  - CV controls the cohesion of block colors which is converging on a single color. Higher voltage = more cohesion and thus their colors will quickly become the same and Photron will look like one color.
  - V/OCT controls the target color for Photron. Blocks will drift towards the target color and hover around it. See the image below for corresponding notes & colors:
 ##### KEYBOARD SHORTCUTS:
- `D+Click` draw with the mouse.

![notes_spectrum](/docs/Notes_Spectrum.png)
- MIDDLE 2 inputs are the X and Y waveforms.

- BOTTOM 3 inputs are triggers:
  - Waveform mode: lines, blocks, or off.
  - Background mode: color, black & white, or black.
  - Draw the pixelated pattern.

#### Photron Panel & Photron Strip

![PhotronPanelStrip](/docs/PhotronPanelStrip.gif)

***NEW!*** *Photron Panel is now resizable*

- `Initialize` will set the colors top to bottom: Purple, Blue, Aqua, and Red.
- `Randomize` will randomize all colors.
##### RIGHT-CLICK MENU:
- `Processing rate` (for those with slower CPUs). Keep in mind, if you slow the processing rate down, it'll help your CPU but the animation will also slow down.
- `Mode` color, black & white, solid color, or strip color.
- `Light` (only effects solid color or strip color mode) one slider to set the pulse speed - in Hz, 0 Hz will be fully on - the second slider will set the hue 0 to 360 degrees.
- `Dark Room Blobs` on or off.
##### KEYBOARD SHORTCUTS:
- `D+Click` draw on Photron Panel.

---

### Polyrhythm Clock

![PolyrhythmClock](/docs/PolyrhythmClock.png)

*A clock featuring three layers of embedded rhythms. It's easy to multiply and/or divide the clock with virtually any rhythm you want using this clock.*

Watch the tutorial:

[![PolyrhythmClock_video](docs/PolyClockRandGates-video.png)](https://youtu.be/v-MF6ziY_bI "Polyrhythm Clock & Rand Gates tutorial")

##### RIGHT-CLICK MENU:
- External Clock Mode:
  - `CV` controls bpm (beats per minute) based on the input voltage using this formula: 120 * 2<sup>V</sup>.
  - `2, 4, 8, 12, 24` `PPQN` controls bpm based on the number pulses per quarter note.
- If the mode is set to any of the `PPQN` modes, the clock will turn on automatically when it receives a pulse. It will also turn off automatically after it times out from not receiving any more pulses.
##### INPUT:
- `RST` resets the clock phases.
- `EXT` is an external clock to control the PolyrhythmClock determined by the External Clock Mode.
- 0v-2v map to 0-24 for the inputs under each knob.
##### BUTTON:
- on or off
##### KNOBS:
- large main knob at the top controls the bpm (beats per minute).
- left-side knobs within each tuplet controls the numerator part of the fraction (or ratio).
- right-side knobs within each tuplet controls the denominator part of the fraction (or ratio).
- middle solid-color knobs within each tuplet controls the probability of outputting a pulse.
##### OUTPUTS:
- first one outputs the current bpm.
- `TUPLET 1` outputs the ratio compared to the bpm.
- `TUPLET 2` outputs the ratio compared to `TUPLET 1`.
- `TUPLET 3` outputs the ratio compared to `TUPLET 2`.

---

### Qubit Crusher

![QubitCrusher](/docs/QubitCrusher.png)

*A bit crusher and downsampler using fractional rates with the ability to modulate bit rate & sample rate, or randomly trigger new bit rates & sample rates.*

##### INPUTS:
- `IN` input signal to be processed.
- `TRG`s both inputs accept gates that trigger random bit rates and/or sample rates.
- Inputs connected to `MOD` knob accept modulation sources (i.e. LFO).
##### KNOBS:
- `BITS` sets bit rate. This is overridden by `TRG` input.
- `SAMP` sets sample rate. This is overriden by `TRG` input.
- `MOD`s set the amount of modulation from modulation signal.
##### OUTPUT:
- `OUT` output signal.

---

### RandGates

![RandGates](/docs/RandGates.png)

*Randomly outputs one of the 4 inputs.*

Watch the tutorial:

[![RandGates_video](docs/PolyClockRandGates-video.png)](https://youtu.be/v-MF6ziY_bI "Polyrhythm Clock & Rand Gates tutorial")

##### INPUTS:
- `TRG` randomizes the output
- `INS` (purple, blue, aqua, red) are any type of input, i.e. gates or ±5 volts.
##### KNOB:
- Weight knob: Gives probability control of the probability to the chosen input. All the way to the right is uniform randomness.
- Weight probability knob: controls the probability of the chosen input.
##### OUTPUT:
- `OUT` outputs either the randomly chosen input as either a pulse or ±5 volts.

---

### RandRoute

![RandRoute](/docs/RandRoute.png)

*Randomly routes one inputs to 4 possible outputs.*

##### INPUTS:
- `TRG` randomizes the output
- `IN` any type of input, i.e. gates or ±5 volts.
##### KNOB:
- Weight knob: Gives probability control of the probability to the chosen output. All the way to the right is uniform randomness.
- Weight probability knob: controls the probability of the chosen output.
##### OUTPUT:
- `OUTS` (purple, blue, aqua, red) randomly chosen to output the input as either a pulse or ±5 volts. If no cable is connected to `TRG` then the `IN` will act as the trigger and output acts like a multinoulli gate only outputting 10 volts or 0 volts.

---

### StochSeq

![StochSeq](/docs/StochSeq.png)

*A sequencer that uses stochastic (probabilistic) patterns. The outputs can be used as gates (triggered based on probability) or as ±5 volts (probability is converted to voltage).*

Click and/or drag to draw your own patterns! Also, click on the different memory banks to recall/store the slider information.

Watch the tutorial:

[![StochSeq_video](docs/StochSeq-video.png)](https://youtu.be/QtNtazAv73M "Stochastic Sequencer tutorial")

##### RIGHT-CLICK MENU:
- Gate mode: gates or triggers.
- V/OCT mode: Independent or Sample and Hold (only changes based on whether gate triggers).
- Volt Offset: ±5V or +10V
- Show or hide slider percentages.
- Enable keyboard shortcuts.
##### KEYBOARD SHORTCUTS:
- `Ctrl+Left` shifts sliders to the left.
- `Ctrl+Right` shifts sliders to the right.
- `Ctrl+Up` shifts sliders up by 5%.
- `Ctrl+Down` shifts sliders down by 5%.
##### INPUTS:
- `CLK` controls timing.
- `RST` resets sequence to beginning of timeline.
- `MEM` CV input to switch between memory banks. A single 1V/oct can control which memory bank to select.
##### KNOBS:
- `LEN` length of the sequence.
- `PATT` selects from preset patterns.
- `SCALE` percentage of the pitch based on the `V/OCT` output and Volt Offset.
- Two knobs control the root note and the scale. The Messiaen modes are based on the wonderful French composer, Olivier Messiaen and his [modes of limited transposition](https://en.wikipedia.org/wiki/Mode_of_limited_transposition) found in his book *The Technique of My Musical Language*.
- Chords like MM7, Mm7, mm7, etc. are based on [seventh chords](https://en.wikipedia.org/wiki/Seventh_chord).
##### BUTTONS:
- `RND` randomizes all probabilities.
- `INV` inverts all probabilities.
- `DIM` cuts the current pattern in half and repeats. Keep clicking this button to continue to diminish the pattern.
##### OUTPUTS:
- `GATE` outputs a pulse based on the probability of the current sequence position. (i.e. a slider at 50% will only trigger a pulse half of the time)
- `V/OCT` outputs pitch based on the slider position and `SPREAD` knob, regardless of probability of the event.
- `INV` outputs invert of `V/OCT`.
- `GATES` outputs correspond to each position in the sequence.
- `NOT` outputs only when `GATE` isn't triggered.

---

### StochSeq4

![StochSeq4](/docs/StochSeq4.png)

*A sequencer that uses four independent stochastic (probabilistic) patterns. The outputs can be used as gates (triggered based on probability) or as ±5 volts (probability is converted to voltage).*

Click and/or drag to draw your own patterns!

Watch the tutorial:

[![StochSeq4_video](docs/StochSeq4-video.png)](https://youtu.be/LBL_VYe_stU "Stoch Seq 4 tutorial")

##### RIGHT-CLICK MENU:
- `MCLK` override: if `MCLK` has a connected cable then it will disable all individual clocks.
- Gate mode: gates or triggers.
- V/OCT mode: Independent or Sample and Hold (only changes based on whether gate triggers).
- Volt Offset: ±5V or +10V
- Show or hide slider percentages.
- Enable keyboard shortcuts.
##### KEYBOARD SHORTCUTS:
- `Ctrl+C` copies focused pattern and length.
- `Ctrl+V` pastes the copied pattern and length to the focused one.
- `Ctrl+Enter` focuses and highlights a single pattern.
- `Ctrl+Left` shifts focused sliders to the left.
- `Ctrl+Right` shifts focused sliders to the right.
- `Ctrl+Up` shifts focused sliders up by 5%.
- `Ctrl+Down` shifts focused sliders down by 5%.
##### INPUTS:
- `MCLK` controls timing of all patterns (overrides all individual clocks).
- `CLK` controls timing of individual patterns.
- `RST` resets sequences to beginning of timeline.
- `RND` gate input randomizes all probabilities.
- `INV` gate input inverts all probabilities.
- `DIM` gate input cuts the current pattern in half and repeats.
##### KNOBS:
- `LEN` length of the individual sequence.
- `PATT` selects from preset patterns.
- `SCALE` percentage of the pitch based on the `V/OCT` output and Volt Offset.
- Two knobs control the root note and the scale just like [StochSeq](#stochseq).
##### BUTTONS:
- Resets sequences to beginning of timeline.
- `RND` randomizes all probabilities.
- `INV` inverts all probabilities.
- `DIM` cuts the current pattern in half and repeats. Keep clicking this button to continue to diminish the pattern.
##### OUTPUTS:
- `GATE` outputs a pulse based on the probability of the current individual sequence position. (i.e. a slider at 50% will only trigger a pulse half of the time)
- `NOT` outputs only when `GATE` isn't triggered.
- `V/OCT` outputs pitch based on the slider position and `SPREAD` knob, regardless of probability of the event.

CHANGELOG.md
v2.2.2 (08-22-2023)
- Added dark skins

v2.2.1 (06-29-2022)
- Added new module PhotronStrip
- Photron modules can be placed adjacent to each other to blend patterns
- Photron can generate pixel-style patterns
- PhotronPanel is resizable

v2.1.1 (02-18-2022)
- Added a memory bank to StochSeq
- Added CV knobs to StochSeqGrid
- External Clock auto start/stop can be overridden

v2.1.0 (01-27-2022)
- Added new module StochSeqGrid
- Updated port UIs
- Added animation to Photron & PhotronPanel for dark room
- Updated volt ranges for StochSeq and StochSeq4

v2.0.2 (12-14-2021)
- Added MCLK override in StochSeq4
- Fixed clock resets in StochSeq and StochSeq4
- Fixed bug in RandRoute multinoulli mode when it's polyphonic
- Updated StochSeq interface

v2.0.1 (12-08-2021)
- Neutrinode node speeds can be extremely slow now
- RandRoute has a multi-bernoulli mode now
- RandRoute & RandGates have a weight probability knob
- RandGates can accept polyphonic triggers and outputs 1 of 4 inputs for each channel
- Added Chromatic scale to quantizer
- Added individual manual URLs to each module

v2.0.0 (11-27-2021)
- Added new module StochSeq4X, an expander to StochSeq4
- Updated StochSeq UI
- Added tooltips to all buttons and knobs
- Updated glowing lights for Rack2 dark room feature
- Neutrinode BPM goes up to 240
- added presets folder

v1.3.6 (10-02-2021)
- Put gates back to 1e-3 on StochSeq and StochSeq4
- Fixed UI bug in sliders that would allow for < 0%

v1.3.5 (9-30-2021)
- Added NOT gate and INV v/oct to StochSeq and all StochSeq4 patterns
- Added OR and XOR gates to StochSeq4 patterns
- StochSeq and StochSeq4 variable width sliders depending on pattern length
- Added keyboard shortcuts to StochSeq and StochSeq4
- more PPQN options to Talea and PolyrhythmClock
- added waveform capabilities to Photron

v1.3.4 (9-17-2021)
- Small changes to interfaces of StochSeq and StochSeq4
- Added Reset input to PolyrhythmClock
- Updated interface to PolyrhythmClock

v1.3.3 (4-8-2021)
- Node movement is now saved in patch with Neutrinode
- Updated lights on StochSeq

v1.3.2 (3-29-2021)
- Added one-shot play mode to Neutrinode
- Updated interface to PolyrhythmClock

v1.3.1 (3-26-2021)
- Added new module RandRoute
- Added new module Collider
- Added volts offset knob for Orbitones
- Bug fix for Orbitones trails
- Updated docs

v1.2.4 (3-05-2021)
- Bug fix in Orbitones
- Added percentages on individual sliders for StochSeq and StochSeq4

v1.2.3 (3-01-2021)
- Bug fixes in Neutrinode and Orbitones

v1.2.2 (1-07-2021)
- Added octave feature to Talea
- Bug fixes
- Updated Docs

v1.2.1 (12-12-2020)
- Added boundary option for particles in Orbitones
- Added input to control target color for Photron
- Bug fixes
- Updated Docs

v1.2.0 (11-30-2020)
- New module Talea (arpeggiator with unique polyrhythm capabilities)
- New module Photron Panel (like photron but panel size)
- Added trails for Orbitones particles
- Bug fixes
- Updated Docs

v1.1.2 (11-12-2020)
- Added right-click menu on Photron to change the processing rates

v1.1.1 (11-08-2020)
- Fixed Photron CPU issue where it was use around 30%, now only ~5%

v1.1.0 (11-02-2020)
- New module Orbitones
- New module Photron
- New module Absorption Spectrum
- Added inputs to StochSeq to control `RND`, `INV`, and `DIM`
- Fixed Polyphony menu in Neutrinode
- Bug fixes
- Updated Docs


v1.0.2 (10-24-2020)
- Added inputs to control `PLAY`, `BPM`, `MOVE`, and root note for Neutrinode
- Added collisions mode (on/off) for Neutrinode nodes
- Added inputs to control `PLAY`, `SPEED`, `RESET`, and root note for Cosmosis
- Added external clock input for PolyrhythmClock


v1.0.1 (10-14-2020)
- Fixed `RESET` input bug for StochSeq4
- Added pitch quantization for both StochSeq and StochSeq4
- Updated Docs


src/Collider.cpp
#include "plugin.hpp"

#define MIN_SHAKE_ENERGY 0.001

struct Collider : Module {
    enum ParamIds {
        SHAKE_PARAM,
        PARTICLES_PARAM,
        CENTER_FREQ_PARAM,
        FREQ_RANGE_PARAM,
        RANDOMIZE_PARAM,
		NUM_PARAMS
	};
	enum InputIds {
        SHAKE_INPUT,
        CENTER_FREQ_INPUT,
        FREQ_RANGE_INPUT,
        VEL_INPUT,
        PARTICLES_INPUT,
		NUM_INPUTS
	};
	enum OutputIds {
        VEL_OUTPUT,
        GATE_OUTPUT,
        VOLT_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightIds {
        PURPLE_LIGHT,
        BLUE_LIGHT,
        AQUA_LIGHT,
        RED_LIGHT,
		NUM_LIGHTS
	};

    dsp::SchmittTrigger shakeTrig;
    dsp::PulseGenerator pulses[16];
    bool gates[16];
    bool isShaking = false;
    float ampLevel = 0.0;
    float centerFreq = 2800; // 2800
    float freqSweep = 1.001;
    float centerVoltage = 0.0; 
    float freqRange = 0.2;
    float notes[3];
    float freqs[3];
    float freqRandomize = 0.0;
    int noteIndex = 0;
    float shakeEnergy = 0.0;
    float velocity = 1.0;
    float systemDecay = 0.9999;
    float soundDecay = 0.85;
    int numOfParticles = 50;
    float percentageObj = numOfParticles * 0.027;
    int currentChannel = 0;
    int channels = 1;
    int checkParams = 0;

    Collider() {
        config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
        configButton(SHAKE_PARAM, "Shake");
        configParam(PARTICLES_PARAM, 1.0, 150.0, 50.0, "Number of Particles");
        configParam(CENTER_FREQ_PARAM, 100.0, 10000.0, 2000.0, "Frequency", " Hz");
        configParam(FREQ_RANGE_PARAM, 0.0, 1.0, 0.2, "Frequency Range", " x");
        configParam(RANDOMIZE_PARAM, 0.0, 1.0, 0.0, "Frequency randomization", " x");

        configInput(SHAKE_INPUT, "Shake");
        configInput(CENTER_FREQ_INPUT, "Center frequency");
        configInput(FREQ_RANGE_INPUT, "Frequency range");
        configInput(VEL_INPUT, "Velocity");
        configInput(PARTICLES_INPUT, "Number of Particles");

        configOutput(VOLT_OUTPUT, "Pitch (V/OCT)");
        configOutput(GATE_OUTPUT, "Gate");
        configOutput(VEL_OUTPUT, "Velocity");

        initNotes(centerFreq);
    }

    void initNotes(float center) {
        centerFreq = center;
        notes[0] = freqToMidi(centerFreq);
        notes[1] = freqToMidi(centerFreq * (1.0 - freqRange));
        notes[2] = freqToMidi(centerFreq * (1.0 + freqRange));

        freqs[0] = centerFreq;
        freqs[1] = centerFreq * (1.0 - freqRange);
        freqs[2] = centerFreq * (1.0 + freqRange);
    }

    json_t *dataToJson() override {
        json_t *rootJ = json_object();
        json_object_set_new(rootJ, "channels", json_integer(channels));

        return rootJ;
    }

    void dataFromJson(json_t *rootJ) override {
        json_t *channelsJ = json_object_get(rootJ, "channels");
        if (channelsJ) channels = json_integer_value(channelsJ);
    }

    void process(const ProcessArgs &args) override {
        if (checkParams == 0) {
            if (params[SHAKE_PARAM].getValue() + inputs[SHAKE_INPUT].getVoltage()) {
                shakeEnergy = velocity;
            }

            if (inputs[VEL_INPUT].isConnected()) {
                float volts = inputs[VEL_INPUT].getVoltage() / 10.0;
                if (volts != velocity) {
                    velocity = volts;
                }
            } else {
                velocity = 1.0;
            }

            freqRandomize = params[RANDOMIZE_PARAM].getValue();

            if (inputs[PARTICLES_INPUT].isConnected()) {
                float cv = inputs[PARTICLES_INPUT].getVoltage() / 10.0;
                cv = cv * cv;
                numOfParticles = (int)rescale(cv, 0.0, 1.0, 1.0, 150.0);
                percentageObj = numOfParticles * 0.027;
            } else if (params[PARTICLES_PARAM].getValue() != numOfParticles) {
                numOfParticles = (int)params[PARTICLES_PARAM].getValue();
                percentageObj = numOfParticles * 0.027;
            }

            if (inputs[CENTER_FREQ_INPUT].isConnected()) {
                if (inputs[CENTER_FREQ_INPUT].getVoltage() != centerVoltage) {
                    centerVoltage = inputs[CENTER_FREQ_INPUT].getVoltage();
                    centerFreq = dsp::FREQ_C4 * pow(2, centerVoltage);
                    initNotes(centerFreq);
                }
            } else if (params[CENTER_FREQ_PARAM].getValue() != centerFreq) {
                initNotes(params[CENTER_FREQ_PARAM].getValue());
            }

            if (inputs[FREQ_RANGE_INPUT].isConnected()) {
                float cv = clamp(inputs[FREQ_RANGE_INPUT].getVoltage() / 10.f, 0.0, 0.95);
                cv = cv * cv; // square it so it's easy to get low values
                if (cv != freqRange) {
                    freqRange = cv;
                    initNotes(centerFreq);
                }
            } else if (params[FREQ_RANGE_PARAM].getValue() != freqRange) {
                freqRange = clamp(params[FREQ_RANGE_PARAM].getValue(), 0.0, 0.95);
                initNotes(centerFreq);
            }
        }
        checkParams = (checkParams + 1) % 4;

        // this algorithm inspired by Perry Cook's Phisem

        if (shakeEnergy > MIN_SHAKE_ENERGY) {
            shakeEnergy *= systemDecay;
            if (randRange(1024.f) < percentageObj) {
                currentChannel = (currentChannel + 1) % channels;
                ampLevel += shakeEnergy;

                pulses[currentChannel].trigger(1e-3f);

                float freq = freqs[noteIndex] * (1.0 + (freqRandomize * randRange(-1.0, 1.0)));
                float volts = freqToVolts(freq);
                noteIndex = (noteIndex + 1) % 3;
                outputs[VOLT_OUTPUT].setVoltage(volts, currentChannel);

                outputs[VEL_OUTPUT].setVoltage(ampLevel * 10.0, currentChannel);
            }
            ampLevel *= soundDecay;

            bool pulse = pulses[currentChannel].process(args.sampleTime);
            outputs[GATE_OUTPUT].setVoltage(pulse ? 10.0 : 0.0, currentChannel);

        } else {
            for (int i = 0; i < channels; i++) {
                bool pulse = pulses[i].process(args.sampleTime);
                outputs[GATE_OUTPUT].setVoltage(pulse ? 10.0 : 0.0, i);
            }
        }

        outputs[VEL_OUTPUT].setChannels(channels);
        outputs[GATE_OUTPUT].setChannels(channels);
        outputs[VOLT_OUTPUT].setChannels(channels);

    }
};

namespace ColliderNS {
    struct ChannelValueItem : MenuItem {
        Collider *module;
        int channels;
        void onAction(const event::Action &e) override {
            module->channels = channels;
        }
    };

    struct ChannelItem : MenuItem {
        Collider *module;
        Menu *createChildMenu() override {
            Menu *menu = new Menu;
            for (int channels = 1; channels <= 16; channels++) {
                ChannelValueItem *item = new ChannelValueItem;
                if (channels == 1)
                    item->text = "Monophonic";
                else
                    item->text = string::f("%d", channels);
                item->rightText = CHECKMARK(module->channels == channels);
                item->module = module;
                item->channels = channels;
                menu->addChild(item);
            }
            return menu;
        }
    };
}

struct ColliderWidget : ModuleWidget {
    ColliderWidget(Collider *module) {
        setModule(module);
        setPanel(createPanel(asset::plugin(pluginInstance, "res/Collider.svg"), asset::plugin(pluginInstance, "res/Collider-dark.svg")));


        // screws
        addChild(createWidget<JeremyScrew>(Vec(17.3, 2)));
        addChild(createWidget<JeremyScrew>(Vec(17.3, box.size.y - 14)));
        addChild(createWidget<JeremyScrew>(Vec(90.7, 2)));
        addChild(createWidget<JeremyScrew>(Vec(90.7, box.size.y - 14)));

        // button
        addParam(createParamCentered<BigButton>(Vec(60, 77.4), module, Collider::SHAKE_PARAM));

        // knobs
        addParam(createParamCentered<BlueKnob>(Vec(29.4, 184.3), module, Collider::CENTER_FREQ_PARAM));
        addParam(createParamCentered<BlueKnob>(Vec(61.5, 184.3), module, Collider::FREQ_RANGE_PARAM));
        addParam(createParamCentered<BlueKnob>(Vec(93.6, 184.3), module, Collider::RANDOMIZE_PARAM));
        addParam(createParamCentered<BlueInvertKnob>(Vec(77.6, 245.8), module, Collider::PARTICLES_PARAM));

        // inputs
        addInput(createInputCentered<PJ301MPort>(Vec(29.4, 106), module, Collider::SHAKE_INPUT));
        addInput(createInputCentered<PJ301MPort>(Vec(29.4, 155.9), module, Collider::CENTER_FREQ_INPUT));
        addInput(createInputCentered<PJ301MPort>(Vec(61.5, 155.9), module, Collider::FREQ_RANGE_INPUT));
        addInput(createInputCentered<PJ301MPort>(Vec(93.6, 155.9), module, Collider::VEL_INPUT));
        addInput(createInputCentered<PJ301MPort>(Vec(45.4, 245.8), module, Collider::PARTICLES_INPUT));

        // outputs
        addOutput(createOutputCentered<PJ301MPort>(Vec(29.4, 300.8), module, Collider::VOLT_OUTPUT));
        addOutput(createOutputCentered<PJ301MPort>(Vec(61.5, 300.8), module, Collider::GATE_OUTPUT));
        addOutput(createOutputCentered<PJ301MPort>(Vec(93.6, 300.8), module, Collider::VEL_OUTPUT));
    }

    void appendContextMenu(Menu *menu) override {
        Collider *module = dynamic_cast<Collider*>(this->module);
        menu->addChild(new MenuEntry);

        ColliderNS::ChannelItem *channelItem = new ColliderNS::ChannelItem;
        channelItem->text = "Polyphony channels";
        channelItem->rightText = string::f("%d", module->channels) + " " + RIGHT_ARROW;
        channelItem->module = module;
        menu->addChild(channelItem);
    }
};

Model *modelCollider = createModel<Collider, ColliderWidget>("Collider");
